[feat] Added Docker examples for standalone output and export output  following best practices

kristiyan-velkov opened the pull request:
## What?

Modernizes the existing `with-docker` example and adds a new `with-docker-export-output` example demonstrating different Next.js deployment strategies with Docker best practices:

1. **`with-docker` (updated)** - Modernized to use App Router, TypeScript, Tailwind CSS v4, and comprehensive Docker best practices. Now includes both Node.js and Bun Dockerfile options, Docker Compose with profiles, BuildKit cache mounts, and extensive documentation.

2. **`with-docker-export-output` (new)** - Demonstrates Next.js static export mode deployment with Docker, offering two serving options: **Nginx** (production-grade) and **serve** package (Node.js-based, simpler setup). Includes optimized Nginx configuration, BuildKit cache mounts, and comprehensive documentation.

Both examples include detailed documentation explaining Docker best practices, implementation decisions, and deployment guidance.

## Why?

The existing `with-docker` example was using the Pages Router and lacked:

- Modern Next.js patterns (App Router, TypeScript)
- Detailed documentation explaining Docker best practices and implementation decisions
- BuildKit cache mounts for optimized build performance
- Bun runtime support as an alternative to Node.js
- Docker Compose configurations for easier local development
- Clear documentation about Node.js image variant choices (slim vs Alpine)
- Examples demonstrating static export mode

These updates serve as comprehensive references for developers who want to understand:

- How to Dockerize Next.js applications for different deployment scenarios
- Why certain choices are made and how to optimize Docker setups for production
- When to use standalone mode vs static export mode
- How to choose between Node.js and Bun runtimes
- How to choose between different web servers (Nginx vs serve) for static sites

## How?

### `with-docker` example (updated)

- **Migrated to App Router** with TypeScript and Tailwind CSS v4
- **Multi-stage Dockerfile** with three stages: dependencies installation, build, and runtime
- **Bun support** via `Dockerfile.bun` with optimized configuration
- **BuildKit cache mounts** for package manager stores (`npm`, `yarn`, `pnpm`, `bun`) and Next.js build cache
- **Security best practices** with non-root user execution (built-in `node` and `bun` users)
- **Docker Compose configuration** with profiles for Node.js (default) and Bun
- **Comprehensive `.dockerignore`** to minimize build context size
- **Detailed README** explaining standalone mode benefits, Node.js image choices, and deployment guidance
- **Deleted** `with-docker-standalone-output` (merged into this example)

### `with-docker-export-output` example (new)

- **Two Dockerfile options:**
  - `Dockerfile` - Nginx-based serving with `nginxinc/nginx-unprivileged` for security
  - `Dockerfile.serve` - Node.js serve package-based serving for simpler deployments
- **Multi-stage builds** for both options with separate dependency, build, and runtime stages
- **BuildKit cache mounts** for package manager stores and Next.js build cache
- **Production Nginx configuration** (`nginx.conf`) with gzip compression, caching headers, and security best practices
- **Docker Compose configuration** supporting both serving options via profiles
- **Comprehensive README** explaining static export mode, trade-offs between Nginx and serve

### Docker best practices implemented (both examples)

| Practice | Description |
|----------|-------------|
| Multi-stage builds | Optimal image size reduction |
| Layer caching | Package files copied first |
| Minimal build context | Comprehensive `.dockerignore` files |
| BuildKit cache mounts | Faster subsequent builds |
| Security hardening | Non-root user execution |
| Corepack | yarn/pnpm version management |
| Package manager auto-detection | npm, yarn, pnpm support |
| Node.js 24.13.0-slim | Clear upgrade guidance |

## Breaking Changes

> **Warning**
> - `with-docker` now uses App Router instead of Pages Router
> - `with-docker-standalone-output` has been removed (merged into `with-docker`)

---

**Note:** After this PR is approved and merged, I plan to open a follow-up PR to update the Next.js documentation with proper links to these Docker examples.

kristiyan-velkov committed:
[feat] Add Docker examples for standalone output and export output following best practices

kristiyan-velkov committed:
fix: update docker standalone example compose.yml and next.config.ts

- Fix PORT environment variable to be explicitly a string in compose.yml
- Remove NextConfig type import from next.config.ts to fix TypeScript error

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'examples/add-docker-examples' of github-private.com:kristiyan-velkov/next.js into examples/add-docker-examples

kristiyan-velkov committed:
chore: remove incorrectly named docker example folder

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

wyattjoh committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

icyJoseph committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

icyJoseph commented:
Hiyo, I wanted to get onto this PR before the holiday break. I think it is a great contribution, though I'd argue that there's some overlap with some of the existing Docker examples. Perhaps, `with-docker-standalone-output` should replace, or enhance further our docker examples?  The other example is a keeper I think, as it is anyway.

kristiyan-velkov commented:
@icyJoseph thanks for the feedback! I agree there’s some overlap. I’d suggest keeping with-docker-standalone-output, as it follows a cleaner, production-ready structure and consistent Dockerfile standards aligned with current best practices. Instead, we could mark the older Docker example as obsolete and update the documentation to point to these new examples as the recommended approach, as noted in the PR. I’m happy to handle the documentation update and cleanup in a follow-up PR, or, if you think it’s better for users, I can update the existing example to match this new structure. I’m currently on holiday, so I have time to react quickly and make any needed adjustments.

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
[feat] update node.js docker image to Node.js LTS 24.12.0

kristiyan-velkov committed:
Merge branch 'examples/add-docker-examples' of github-private.com:kristiyan-velkov/next.js into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov commented:
Hi @icyJoseph, hope you’re doing well. Just a gentle follow-up on this from about a month ago — did you get a chance to take a look at the suggestion regarding consolidating or updating the Docker examples? No rush at all, just wanted to check in and see how you’d like to proceed. Happy to help with any updates or follow-ups if needed.

icyJoseph commented:
@kristiyan-velkov hi there, I had to take a short pause on work, but I just got back this week. I'll get back to this ASAP, sorry for the wait!

icyJoseph commented:
Ok, so the export example is 100% I'd want to merge right now I think

The standalone example, I think if could instead modify with-docker with those changes, **_that'd be great_**. The example name is used quite a lot. During the CVE week, I saw many referring back to it and how it had saved them because it doesn't use a root user. Thanks once again for the PR, and having waited this long.

kris-velkov committed:
feat(examples): modernize with-docker with App Router and improved Docker patterns

kris-velkov committed:
Merge branch 'examples/add-docker-examples' of github-private.com:kristiyan-velkov/next.js into examples/add-docker-examples

kristiyan-velkov commented:
> Ok, so the export example is 100% I'd want to merge right now I think
> 
> The standalone example, I think if could instead modify with-docker with those changes, **_that'd be great_**. The example name is used quite a lot. During the CVE week, I saw many referring back to it and how it had saved them because it doesn't use a root user. Thanks once again for the PR, and having waited this long.

Thanks for the feedback and for your patience @icyJoseph ! I've updated the PR as you suggested:

**Changes made:**

- Merged with-docker-standalone-output into with-docker - The existing with-docker example now includes all the improvements from the standalone example while preserving its well-known name
- Updated all node.js images tags to LTS version of node.js 24.13.0.
- Deleted with-docker-standalone-output - Since its content is now in with-docker, this is no longer needed
- with-docker-export-output - Ready to merge as-is
- What's updated in with-docker:
- Migrated from Pages Router to App Router with TypeScript
- Added Bun support via Dockerfile.bun with Docker Compose profiles
- Added BuildKit cache mounts for faster builds (npm, yarn, pnpm, bun)
- Continues using non-root users (node for Node.js, bun for Bun) - the security feature you mentioned
- Updated to Node.js 24.13.0-slim with comprehensive documentation
- Added Docker Compose configuration with Node.js (default) and Bun profiles
- Comprehensive README explaining Docker best practices and deployment options

The with-docker name and its security-focused approach (non-root user) are preserved, just modernized with current best practices. Happy to make any additional adjustments if needed! Pleasure to work with you!

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

icyJoseph commented on examples/with-docker/README.md:
```diff
@@ -1,83 +1,187 @@
-This will build the project as a standalone app inside the Docker image.
+The standalone output mode creates a minimal, self-contained production build optimized for containerized deployments. When enabled, Next.js generates a `.next/standalone` directory containing only the essential files needed to run your application, excluding unnecessary dependencies and files. This results in significantly smaller Docker images and faster container startup times.
 
-## Deploying to Google Cloud Run
```
Maybe we want to keep this instruction bit?

kristiyan-velkov replied:
Good catch! 

I've restored both:
- Google Cloud Run deployment instructions - Added back the step-by-step guide for deploying to GCR
- "In existing projects" section - Added instructions for adding Docker support to existing Next.js projects (copy Dockerfile, .dockerignore, and add output: "standalone" to config)
Thanks for the review!

icyJoseph commented on examples/with-docker/README.md:
```diff
@@ -1,83 +1,187 @@
+---
 
-The `pages/api` directory is mapped to `/api/*`. Files in this directory are treated as [API routes](https://nextjs.org/docs/api-routes/introduction) instead of React pages.
+Made with ❤️ by [Kristiyan Velkov](https://kristiyanvelkov.com/)
```
Let's keep these author agnostic

kristiyan-velkov replied:
Ok, done!

kris-velkov committed:
docs(with-docker): restore deployment instructions and add existing projects guide

kris-velkov committed:
Merge branch 'examples/add-docker-examples' of github-private.com:kristiyan-velkov/next.js into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

icyJoseph commented on examples/with-docker-export-output/README.md:
```diff
@@ -0,0 +1,243 @@
+- [Front-end Production Dockerfiles](https://github.com/kristiyan-velkov/frontend-prod-dockerfiles) - Production-ready Dockerfiles for React.js, Angular, Vue.js, Next.js, Nuxt.js and more front-end applications.
+
+## Author
+
```
Here too - I don't think we have any examples referencing a given author? All of them are agnostic

I also noticed some metadata further down and such, let's try to have the same UI as the previous example (with-docker) did and align this with-docker-export-output to that too

kristiyan-velkov replied:
Fixed! Removed all personal attribution (author sections, social links, personal repo links) from both READMEs and page.tsx files. Also added Docker Compose profile for the serve variant. Both examples now align with standard Next.js example conventions.

icyJoseph commented on examples/with-docker-export-output/nginx.conf:
```diff
@@ -0,0 +1,57 @@
+
+        # Handle Next.js client-side routing
+        location / {
+            try_files $uri $uri/ /index.html;
```
Is this correct? https://nextjs.org/docs/app/guides/static-exports#deploying shows:

```nginx
  location / {
      try_files $uri $uri.html $uri/ =404;
  }
 
  # This is necessary when `trailingSlash: false`.
  # You can omit this when `trailingSlash: true`.
  location /blog/ {
      rewrite ^/blog/(.*)$ /blog/$1.html break;
  }
```

kristiyan-velkov replied:
Good point! I've updated nginx.conf to match the Next.js static export docs.

icyJoseph commented on examples/with-docker-export-output/compose.yml:
```diff
@@ -0,0 +1,25 @@
+    restart: unless-stopped
+
+  nextjs-static-export-with-serve:
+    build:
```
Should this one have a profile?

kristiyan-velkov replied:
I've added a serve profile to the nextjs-static-export-with-serve service.
Now it matches the with-docker pattern:
- docker compose up → runs Nginx (default/recommended)
- docker compose --profile serve up → runs serve variant
Updated the README commands accordingly.

kris-velkov committed:
chore(examples): remove personal attribution from Docker examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

icyJoseph commented:
Taking a look this week once again. Sorry it has been dragging a bit!

icyJoseph commented on examples/with-docker-export-output/app/layout.tsx:
```diff
@@ -0,0 +1,63 @@
+    "Node.js",
+    "Nginx",
+  ],
+  authors: [
```
here and in the with-docker layout we should also drop these

otherwise, I don't think I have anymore observations

kristiyan-velkov replied:
Good point — I’ve dropped these.
Thank you for your support and thoughtful comments. It’s a pleasure working with you.

kris-velkov committed:
chore(examples): align with-docker examples to Next.js standards

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

icyJoseph commented on examples/with-docker-export-output/app/layout.tsx:
```diff
@@ -0,0 +1,57 @@
+export const metadata: Metadata = {
+  title: "Next.js Docker Example - Standalone & Static Export",
+  description:
+    "A production-ready example demonstrating how to Dockerize Next.js applications using standalone mode and static export. Created by Kristiyan Velkov, Docker Captain.",
```
```suggestion
    "A production-ready example demonstrating how to Dockerize Next.js applications using standalone mode and static export.",
```

icyJoseph commented on examples/with-docker/app/layout.tsx:
```diff
@@ -0,0 +1,61 @@
+export const metadata: Metadata = {
+  title: "Next.js Docker Example - Standalone Mode",
+  description:
+    "A production-ready example demonstrating how to Dockerize Next.js applications using standalone mode. Created by Kristiyan Velkov, Docker Captain.",
```
```suggestion
    "A production-ready example demonstrating how to Dockerize Next.js applications using standalone mode.",
```

icyJoseph commented on examples/with-docker/app/layout.tsx:
```diff
@@ -0,0 +1,61 @@
+      name: "Kristiyan Velkov",
+      url: "https://www.linkedin.com/in/kristiyan-velkov-763130b3/",
+    },
+  ],
```
```suggestion
```

kris-velkov committed:
remove author info Kristiyan Velkov

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

icyJoseph commented:
One last thing... I was trying out the examples locally and noticed that, the `--profile` approach doesn't work as intended for mutually exclusive services. Docker Compose profiles are additive - running `docker compose --profile bun up` starts both the default service AND the bun service, causing a port conflict on 3000.

I think we should go back to how you had it:

- remove `profiles: [bun]` and update README:

```bash
# Run with Node.js
docker compose up nextjs-standalone --build

# OR run with Bun
docker compose up nextjs-standalone-with-bun --build
```

- remove `profiles: [serve]` and update README:

```bash
# Run with Nginx (recommended)
docker compose up nextjs-static-export --build

# OR run with serve
docker compose up nextjs-static-export-with-serve --build
```

Sorry for that, this is the very last thing I think, everything else is in place.

kris-velkov committed:
docker examples: run one service via name to avoid profile port conflict

kristiyan-velkov commented:
> One last thing... I was trying out the examples locally and noticed that, the `--profile` approach doesn't work as intended for mutually exclusive services. Docker Compose profiles are additive - running `docker compose --profile bun up` starts both the default service AND the bun service, causing a port conflict on 3000.
> 
> I think we should go back to how you had it:
> 
> * remove `profiles: [bun]` and update README:
> 
> ```shell
> # Run with Node.js
> docker compose up nextjs-standalone --build
> 
> # OR run with Bun
> docker compose up nextjs-standalone-with-bun --build
> ```
> 
> * remove `profiles: [serve]` and update README:
> 
> ```shell
> # Run with Nginx (recommended)
> docker compose up nextjs-static-export --build
> 
> # OR run with serve
> docker compose up nextjs-static-export-with-serve --build
> ```
> 
> Sorry for that, this is the very last thing I think, everything else is in place.

No problem  at all 👍

Profiles removed, README updated, commands now use explicit services.

Changes are done. Pleasure to work with you!

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

kristiyan-velkov committed:
Merge branch 'canary' into examples/add-docker-examples

icyJoseph reviewed (approved):
Only a minor nit. I think we are good to go. 

nit: the `serve` example routes all to `/` which means no 404's but that can be fixed or explained later.

icyJoseph committed:
Merge branch 'canary' into examples/add-docker-examples

icyJoseph commented:
Let's follow up with docs updates and such! Good stuff!

kristiyan-velkov commented:
Hi @icyJoseph. 

Thank you for merging the PR — I truly appreciate the support and the quick turnaround.

I’ll move forward with the documentation updates right away and make sure everything is properly reflected in the docs as soon as possible.

Thanks again for the collaboration! Pleasure to work with you! 

Best regards,
Kristiyan Velkov
